### ISSUE TITLE
runtime changes not being applied

### ISSUE THREAD

**Issue by gambol99** [Mashape/kong#1509]
### Summary

Very odd one, I have Kong running in a AWS autoscaling group using docker container. At the moment, I only have a single instances in the cluster. I can make changes via the API, I can call the Admin API and see the changes have been applied. But when I call actual proxy, the configuration does looks like it's been applied. Example, creating and API, view its config and everything looks fine, by call the API and it throws a API not found. Running kong reload does not work, but if i kong restart or simply kill the container, the configuration has been applied.
### Steps To Reproduce
1. Simply make a change to API and check it has taken place
### Additional Details & Logs
- Kong version: 0.8.3
- Operating System: CoreOS 
- Kong configuration 

``` YAML
#
# Kong Configuration
#
proxy_listen: "0.0.0.0:8000"
proxy_listen_ssl: "0.0.0.0:8443"
admin_api_port: "0.0.0.0:8001"
send_anonymous_reports: false
memory_cache_size: 128
cluster_listen: "0.0.0.0:7946"
cluster_listen_rpc: "127.0.0.1:7373"
cluster:
  ttl_on_failure: 120
database: postgres
postgres:
  host: "HOSTNAME"
  port: 5432
  database: DBNAME
  user: USERNAME
  password: PASSWORD
ssl_cert_path: /etc/secrets/proxy.pem
ssl_key_path: /etc/secrets/proxy-key.pem
nginx: |
  {{user}}
  worker_processes auto;
  error_log /dev/stderr error;
  daemon off;
  worker_rlimit_nofile {{auto_worker_rlimit_nofile}};
  env KONG_CONF;
  env PATH;
  events {
    worker_connections {{auto_worker_connections}};
    multi_accept on;
  }
  http {
    resolver {{dns_resolver}} ipv6=off;
    charset UTF-8;

    include /etc/nginx/conf/logging-formats.conf;
    include /etc/nginx/conf/logging.conf;

    # Timeouts
    keepalive_timeout 60s;
    client_header_timeout 60s;
    client_body_timeout 60s;
    send_timeout 60s;

    # Proxy settings
    proxy_buffer_size 128k;
    proxy_buffers 4 256k;
    proxy_busy_buffers_size 256k;
    proxy_ssl_server_name on;

    # IP Address
    real_ip_recursive on;

    # Other Settings
    client_max_body_size 0;
    underscores_in_headers on;
    reset_timedout_connection on;
    tcp_nopush on;

    ################################################
    #  The following code is required to run Kong  #
    # Please be careful if you'd like to change it #
    ################################################
    # Lua Settings
    lua_package_path ';;';
    lua_code_cache on;
    lua_max_running_timers 4096;
    lua_max_pending_timers 16384;
    lua_shared_dict reports_locks 100k;
    lua_shared_dict cluster_locks 100k;
    lua_shared_dict cluster_autojoin_locks 100k;
    lua_shared_dict cache {{memory_cache_size}}m;
    lua_shared_dict cassandra 1m;
    lua_shared_dict cassandra_prepared 5m;
    lua_socket_log_errors off;
    {{lua_ssl_trusted_certificate}}
    init_by_lua_block {
      kong = require "kong"
      kong.init()
    }
    init_worker_by_lua_block {
      kong.init_worker()
    }
    server {
      server_name _;
      listen {{proxy_listen}} proxy_protocol;
      listen {{proxy_listen_ssl}} ssl proxy_protocol;
      real_ip_header proxy_protocol;
      set_real_ip_from 10.0.0.0/8;
      real_ip_recursive on;
      ssl_certificate_by_lua_block {
        kong.ssl_certificate()
      }
      ssl_certificate {{ssl_cert}};
      ssl_certificate_key {{ssl_key}};
      ssl_protocols TLSv1.1 TLSv1.2;
      location / {
        default_type 'text/plain';
        # These properties will be used later by proxy_pass
        set $upstream_host nil;
        set $upstream_url nil;
        # Authenticate the user and load the API info
        access_by_lua_block {
          kong.access()
        }
        proxy_set_header X-Real-IP $proxy_protocol_addr;
        proxy_set_header X-Forwarded-For $proxy_protocol_addr;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header Host $upstream_host;
        proxy_pass $upstream_url;
        proxy_pass_header Server;
        # Add additional response headers
        header_filter_by_lua_block {
          kong.header_filter()
        }
        # Change the response body
        body_filter_by_lua_block {
          kong.body_filter()
        }
        # Log the request
        log_by_lua_block {
          kong.log()
        }
      }
      location /robots.txt {
        return 200 'User-agent: *\nDisallow: /';
      }
      error_page 500 502 503 504 /50x;
      location = /50x {
        internal;
        content_by_lua_block {
          require("kong.core.error_handlers")(ngx)
        }
      }
    }
    server {
      listen {{admin_api_listen}} ssl proxy_protocol;
      real_ip_header proxy_protocol;
      set_real_ip_from 10.0.0.0/8;
      client_max_body_size 10m;
      client_body_buffer_size 10m;
      ssl_certificate {{ssl_cert}};
      ssl_certificate_key {{ssl_key}};
      ssl_protocols TLSv1.1 TLSv1.2;
      location / {
        include /etc/nginx/conf/auth.conf;
        default_type application/json;
        content_by_lua_block {
          ngx.header["Access-Control-Allow-Origin"] = "*"
          if ngx.req.get_method() == "OPTIONS" then
            ngx.header["Access-Control-Allow-Methods"] = "GET,HEAD,PUT,PATCH,POST,DELETE"
            ngx.header["Access-Control-Allow-Headers"] = "Content-Type"
            ngx.exit(204)
          end
          local lapis = require "lapis"
          lapis.serve "kong.api.app"
        }
      }
      location /nginx_status {
        internal;
        access_log off;
        stub_status;
      }
      location /robots.txt {
        return 200 'User-agent: *\nDisallow: /';
      }
    }
  }
```

**Comment by gambol99** [comment]
thought it might have been db related somehow ... so rebuilt the RDS instance at version v9.4.7 from v9.5 .. Still no luck no ... If i add an API is appears, if I delete it, it remains there until I restart.

**Comment by thefosk** [comment]
@gambol99 while I look into this, in the meanwhile, can you try to use a new 0.9 RC to see if this is fixed?

**Comment by gambol99** [comment]
@thefosk ... i'll look at 0.9rc

**Comment by patrickpierson** [comment]
I have seen this same issue on 0.8.3.  All of our api changes are done via https://github.com/PGBI/kong-dashboard however to verify I also made updates via curl and on 0.8.3 I had the exact same issues as @gambol99.  Updating to 0.9.0 however fixed that issue slightly.  I am able to apply new changes without restarting the container however when deleting for example a key auth value (apikey) for a consumer that key auth value (apikey) continues to exist until restart of the container.

long story short
0.8.3 -> all changes made require a restart
0.9.0 -> deletions require a restart

full disclosure I did change from cassandra on 0.8.3 to postgresql on 0.9.0 but I have no input on if that affects this at all.

**Comment by thefosk** [comment]
If you check the "/cluster" endpoint, do you see all the Kong nodes
properly clustered together?


 *Marco Palladino *| CTO @ Mashape | mashape.com | +1 (415) 361-3858

**Comment by patrickpierson** [comment]
that explains it.  thanks @thefosk

**Comment by dheerajdwivedi** [comment]
Facing the same issue in kong/0.9.2
Admin changes are not reflected unless I restart kong.

Output of admin:8001/cluster is
 `{"data":[
{"address":"*:7946","name":"kong01_0.0.0.0:7946_asasas","status":"alive"},{"address":"*:7946","name":"kong02_0.0.0.0:7946_aasasad","status":"alive"}],"total":2}`

Using Cassandra DB and I can see the data being updated.

**Comment by thefosk** [comment]
@dheerajdwivedi I have updated the clustering docs: https://getkong.org/docs/latest/clustering/

Can you check that `kong cluster members` returns the right number of members, and that `kong cluster reachability` can successfully check that the nodes are able to communicate together on both TCP and UDP?

**Comment by dheerajdwivedi** [comment]
Output of `kong cluster reachability`
`Total members: 2, live members: 2
Starting reachability test...
Successfully contacted all live nodes`

There is no problem with clustering it seems.
Problem persist even with single node.
After updating the data via admin api's(8001), the data is not being refreshed in proxy automatically. It requires `kong restart` for changes to reflect which is not the expected behavior.
Example:
Updated the api's upstream url but the request is still going to older upstream url.

**Comment by pamiel** [comment]
I'm just facing the same issue as the one originally descibed by @gambol99 , but after driving me crazy, I just solved the problem! Let me share with you:

Issue is: API is correctly created when calling the admin endpoint, it can immediately be viewed from the <host>:8001/apis ... but no way to call it using the proxy endpoint: the error {"message":"no API found with those values"} is raised. The only way is to restart Kong, and the API becomes “magically” accessible in the proxy endpoint…

It is very easy to reproduce (just launch cassandra, launch Kong, declare the API and call it) and I experienced it the same way using 0.9.6, 0.9.8 and 0.10.0rc... but my environment was a bit special: this is on a recompiled Kong, packaged into a Docker image.

Serf executable is not in the regular path, but logs available via the “docker logs” says that it is found (thanks to the updated PATH, I guess).
Outputs of “kong cluster members” and “kong cluster reachability” command looks ok as well…
 
However, a call to <host>:8001/cluster returns a {"message":"An unexpected error occurred"} ! Digging into the nginx log files, I finally found an error:
2017/01/27 17:56:56 [error] 83#0: *8 [lua] hooks.lua:156: member_join(): sh: serf: not found, context: ngx.timer

Despite the presence of serf in the PATH, it looks hooks.lua is not able to access it.

The solution is “just” to add the serf_path variable in the kong.conf file, and it works !
Note that serf_path is not very documented (only in the "compile source" part... my fault having missed this line !), but it is not present in the kong.conf.default file => maybe it should be documented there as well, no ?

Hoping it could help someone else !
